In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('/content/diabetes.csv')

In [3]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [4]:
# Replace 0 with NaN in all columns except 'Outcome'
df.loc[:, df.columns != 'Outcome'] = df.loc[:, df.columns != 'Outcome'].replace(0, np.nan)

<ipython-input-4-d2c510afc6e4>:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[ 6.  1.  8.  1. nan  5.  3. 10.  2.  8.  4. 10. 10.  1.  5.  7. nan  7.
  1.  1.  3.  8.  7.  9. 11. 10.  7.  1. 13.  5.  5.  3.  3.  6. 10.  4.
 11.  9.  2.  4.  3.  7.  7.  9.  7. nan  1.  2.  7.  7.  1.  1.  5.  8.
  7.  1.  7. nan nan nan  2.  8.  5.  2.  7.  5. nan  2.  1.  4.  2.  5.
 13.  4.  1.  1.  7.  5. nan  2.  3.  2.  7. nan  5.  2. 13.  2. 15.  1.
  1.  4.  7.  4.  2.  6.  2.  1.  6.  1.  1.  1. nan  1.  2.  1.  1.  4.
  3. nan  3.  8.  1.  4.  7.  4.  5.  5.  4.  4. nan  6.  2.  5. nan  1.
  3.  1.  1. nan  4.  9.  3.  8.  2.  2. nan nan nan  5.  3.  5.  2. 10.
  4. nan  9.  2.  5.  2.  1.  4.  9.  1.  8.  7.  2.  1.  2. 17.  4.  7.
 nan  2. nan  6.  3.  4.  4.  3.  6.  6.  2.  1.  2.  8.  6. nan  5.  5.
  6. nan  1.  5.  4.  7.  8.  1.  8.  5.  3.  9.  7. 11.  8.  5.  1.  3.
  4.  4. nan  1. nan  2.  6.  5.  8.  5.  1. 

In [5]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0          6.0    148.0           72.0           35.0      NaN  33.6   
1          1.0     85.0           66.0           29.0      NaN  26.6   
2          8.0    183.0           64.0            NaN      NaN  23.3   
3          1.0     89.0           66.0           23.0     94.0  28.1   
4          NaN    137.0           40.0           35.0    168.0  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               657 non-null    float64
 1   Glucose                   763 non-null    float64
 2   BloodPressure             733 non-null    float64
 3   SkinThickness             541 non-null    float64
 4   Insulin                   394 non-null    float64
 5   BMI                       757 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(7), int64(2)
memory usage: 54.1 KB


In [7]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   657.000000  763.000000     733.000000     541.000000  394.000000   
mean      4.494673  121.686763      72.405184      29.153420  155.548223   
std       3.217291   30.535641      12.382158      10.476982  118.775855   
min       1.000000   44.000000      24.000000       7.000000   14.000000   
25%       2.000000   99.000000      64.000000      22.000000   76.250000   
50%       4.000000  117.000000      72.000000      29.000000  125.000000   
75%       7.000000  141.000000      80.000000      36.000000  190.000000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  757.000000                768.000000  768.000000  768.000000  
mean    32.457464                  0.471876   33.240885    0.348958  
std      6.924988                  0.331329   11.760232    0.476951  
min     18.200000                  0.078000   21.000000    0.000000  
25%     27.500000                  0.243750   24.000000    0.000000  
50%     32.300000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [8]:
df.isnull().sum()

Pregnancies                 111
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

#Handle Null Values

In [9]:
# Replace NaN with mode for specific columns
mode_columns = ['Pregnancies', 'BloodPressure', 'SkinThickness', 'Glucose', 'Insulin']
for col in mode_columns:
    mode_value = df[col].mode()[0]  # mode() returns a Series, [0] gives the most frequent value
    df[col].fillna(mode_value, inplace=True)

# Replace NaN with mean for BMI
df['BMI'].fillna(df['BMI'].mean(), inplace=True)


<ipython-input-9-91bdb01a29d8>:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[col].fillna(mode_value, inplace=True)
<ipython-input-9-91bdb01a29d8>:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.meth

In [10]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [11]:
df['Outcome'].unique()

array([1, 0])

#Spliting into X and Y

In [12]:
x= df.drop('Outcome',axis = 1)
y = df['Outcome']

#Split into train and test dataset

In [13]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=.2,random_state=42)

#Scaling

In [14]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
x_train = sc.fit_transform(x_train)
x_test  = sc.fit_transform(x_test)

#Logistic Regression

In [15]:
from sklearn.linear_model import LogisticRegression
lr =LogisticRegression()
lr.fit(x_train,y_train)

LogisticRegression()

In [16]:
y_pred = lr.predict(x_test)

In [17]:
y_test

668    0
324    0
624    0
690    0
473    0
      ..
355    1
534    0
344    0
296    1
462    0
Name: Outcome, Length: 154, dtype: int64

In [18]:
y_pred

array([0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0,
       0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0,
       0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1,
       0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0])

In [19]:
#Confusion Matrix
from sklearn.metrics import confusion_matrix,accuracy_score,f1_score,precision_score,recall_score
confusion_matrix(y_test,y_pred)

array([[85, 14],
       [20, 35]])

85 - True Positive

35 - True Negative

14 - False positive

20 - False Negative

In [20]:
print(accuracy_score(y_test,y_pred))
print(precision_score(y_test,y_pred))
print(recall_score(y_test,y_pred))
print(f1_score(y_test,y_pred))

0.7792207792207793
0.7142857142857143
0.6363636363636364
0.6730769230769231


Accuracy: 0.779 — about 77.9% of total predictions were correct.

Precision: 0.714 — when the model predicted positive, it was correct 71.4% of the time.

Recall (Sensitivity): 0.636 — the model correctly identified 63.6% of all actual positives.

F1 Score: 0.673 — the harmonic mean of precision and recall, giving a balance between the two.

Good overall accuracy (~78%) and precision, which suggests the model is reliable

#KNN

In [21]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=7)
knn.fit(x_train,y_train)

KNeighborsClassifier(n_neighbors=7)

In [22]:
y_pred_knn = knn.predict(x_test)
accuracy_score(y_test,y_pred_knn)

0.6818181818181818

KNN have a Accuracy Score of 68% which is less than LogisticRegression

#SVC

In [23]:
from sklearn.svm import SVC
svm = SVC()
svm.fit(x_train,y_train)

SVC()

In [24]:
y_pred_svm = svm.predict(x_test)
accuracy_score(y_test,y_pred_svm)

0.7467532467532467

SVC have an accuracy of 74.67%

Among the above LogisticRegression have the best performace